### PR TITLE
[FLINK-24861][connector][jdbc] Fix false cache lookup for empty data

### PR DESCRIPTION
**Pull request by gaurav726**
## What is the purpose of the change

Ideally, in case of cache miss for a key, or with null value fetch for key, key shouldn't be cached, however empty result query should also be in the scope, so including an extra optional parameter to configure same

## Brief change log

empty check on row list will make sure for database read for empty data for the key

## Verifying this change

This change is a trivial check / code fix without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no**/ don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by hililiwei** [comment]
look good

**Comment by gaurav726** [comment]
Hi @rmetzger , I have made the changes as discussed on JIRA ticket, somehow azure pipeline not triggered after push, tried closing PR as well

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java
Comment (wenlong88) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java:178]: you need to a new case instead of override the default value

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java
Comment (wenlong88) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java:152]: i am afraid that you should not just ignore the empty values, instead you should remove the keyRow from the cache if exlucdeEmptyQueryResult is true, and collect is still needed.

**Review by wenlong88 — COMMENTED** [review]
@gaurav726  thanks for the contribution, I left some comments

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java
Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java:154]: Hi @wenlong88 , changes has been done, please review

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java
Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java]: you mean new test case,right?, I have added another one,

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java
Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java:154]: Hi @wenlong88 , pipeline got succeed, can you please review and merge

Comment (wenlong88) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java:154]: hi, you need to add a test cover the cache changes, refer to JdbcRowDataLookupFunctionTest#testEval, on the other hands, I think it is more clearer if you do not put empty to cache in eval instead of put to cache and then invalidate.

Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java:154]: cool, got it, making changes

Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: Hi @wenlong88, code changes has been done, but for test case, I have a doubt, If you look at JdbcRowDataLookupFunctionTest#testEval, it test for collector, but for my use case, output will never get collected in collector, Also there is no getter for cache for which i can assert, should I put a getter for cache in this class, can you please help me with test case for it

Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: Hi @wenlong88 , did you checked?

Comment (wenlong88) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: you should verify that the empty value is not in the cache, will not affect other eval. for example, eval for the time with empty result, update the db, eval another time, the second eval could get the latest result .

Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: @wenlong88 , But for this, I need getter for `cache` instance in JdbcRowDataLookupFunction, should I add it? I can't test with collector object, correct?

Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: @wenlong88  added test cases, please review

**Comment by wenlong88** [comment]
LGTM for now, cc @wuchong

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java
Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: @wenlong88 can we resolve conversations now? Also @wuchong will get it merge?

Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: @wenlong88  anything else is remaining in this?

Comment (wenlong88) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: @gaurav726 the pr looks good for me now, but you still need to wait for Committers to join the review.

Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: cool, thanks for the update @wenlong88

Comment (gaurav726) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataLookupFunction.java]: @wuchong can you please review my PR.

**Comment by wuchong** [comment]
Thanks @gaurav726 for the great work. The implementation looks good to me. However, I have some concerns about the option key. From my point of view, this issue wants to disable caching missing key (same problem with FLINK-21415). Therefore, this option key would be better to belong to `lookup.cache.xx`. How about naming it `lookup.cache.caching-missing-key=true` which looks more intuitive?

**Comment by gaurav726** [comment]
Hi @wuchong , thanks for reviewing my PR, making changes

**Comment by gaurav726** [comment]
Hi @wuchong , changes has been done, please review, pipeline is running, will update once success

**Comment by gaurav726** [comment]
Hi @wuchong, pipeline got succeed, please review and merge

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**
